### PR TITLE
Rewrite `Parser::collect_tokens`

### PR DESCRIPTION
**Pull request by Aaron1011**
The previous implementation did not work when called on an opening
delimiter, or when called re-entrantly from the same `TokenCursor` stack
depth.

I'm not sure how to test this apart from https://github.com/rust-lang/rust/pull/72287

### CODE REVIEW

File: src/librustc_parse/parser/mod.rs [hunk 1]
```diff
@@ -184,9 +179,22 @@ impl TokenCursor {
...
+            }
+
+            if let Some(collecting) = &mut self.collecting {
+                if collecting.depth == self.stack.len() {
+                    debug!(
+                        "TokenCursor::next():  collected {:?} at depth {:?}",
+                        tree,
+                        self.stack.len()
+                    );
+                    collecting.buf.push(tree.clone().into())
```
Comment (petrochenkov) [inline src/librustc_parse/parser/mod.rs:196]: Looks like we are losing jointness when collecting tokens.
This is pre-existing, but should be easy to fix with
```diff
-} else if let Some(tree) = self.frame.tree_cursor.next() {
+} else if let Some(tree) = self.frame.tree_cursor.next_with_joint() {
```
above.

File: src/librustc_parse/parser/mod.rs [hunk 2]
```diff
@@ -1105,65 +1115,95 @@ impl<'a> Parser<'a> {
...
+                // will be duplicates of the tokens nested inside the `TokenTree::Delimited`.
+                // Therefore, we set our recording depth to the *previous* frame. This allows
+                // us to record a sequence like: `(foo).bar()`: the `(foo)` will be recored
+                // as our initial `cur_token`, while the `.bar()` will be recored after we
+                // pop the `(foo)` frame.
+                self.token_cursor.stack.len() - 1
+            } else {
+                self.token_cursor.stack.len()
+            };
+        self.token_cursor.collecting = Some(Collecting { buf: tokens, depth: prev_depth });
```
Comment (petrochenkov) [inline src/librustc_parse/parser/mod.rs:1169]:
```suggestion
        let prev_collecting = self.token_cursor.collecting.replace(Collecting { buf: tokens, depth: prev_depth });
```

**Comment by Aaron1011** [comment]
@petrochenkov: Updated

**Comment by petrochenkov** [comment]
r=me after squashing commits

**Pull request merged**